kubespawner/objects.py

```diff
@@ -566,3 +610,103 @@ def make_ingress(
     )
 
     return endpoint, service, ingress
+
+
+def make_secret(
+    name,
+    env=None,
+    labels=None,
+    annotations=None,
+):
+    """
+    Make a k8s secret specification for mount ssl credentials for running a user notebook.
+
+    Parameters
+    ----------
+    name:
+        Name of the secret. Must be unique within the namespace the object is
+        going to be created in.
+    env:
+        Dictionary of environment variables.
+    labels:
+        Labels to add to the spawned pod.
+    annotations:
+        Annotations to add to the spawned pod.
+
+    """
+
+    if not ssl_enabled(env):
+        return None
+
+    secret = V1Secret()
+    secret.kind = "Secret"
+    secret.api_version = "v1"
+    secret.metadata = V1ObjectMeta()
+    secret.metadata.name = name
+    secret.metadata.annotations = (annotations or {}).copy()
+    secret.metadata.labels = (labels or {}).copy()
+
+    secret.data = {}
+    with open(env['JUPYTERHUB_SSL_KEYFILE'], 'r') as file:
+        encoded = base64.b64encode(file.read().encode("utf-8"))
+        secret.data['key'] = encoded.decode("utf-8")
+    with open(env['JUPYTERHUB_SSL_CERTFILE'], 'r') as file:
+        encoded = base64.b64encode(file.read().encode("utf-8"))
+        secret.data['crt'] = encoded.decode("utf-8")
+    with open(env['JUPYTERHUB_SSL_CLIENT_CA'], 'r') as file:
+        encoded = base64.b64encode(file.read().encode("utf-8"))
+        secret.data["notebooks-ca_trust.crt"] = encoded.decode("utf-8")
+
+    folder = os.path.dirname(env['JUPYTERHUB_SSL_CLIENT_CA'])
+    with open(folder+'/hub-ca_trust.crt', 'r') as file:
+        encoded = base64.b64encode(file.read().encode("utf-8"))
+        secret.data["notebooks-ca_trust.crt"] = secret.data["notebooks-ca_trust.crt"] + encoded.decode("utf-8")
+
+    return secret
+
+
+def make_service(
```

Comment:
I'm a little torn about this. JupyterHub's internal_ssl supports using 'IP' as part of alt_names, but in Kubernetes we do not know the IP until after the pod is created - at which point we can not add a secret there...

There is already code here for creating a new Service per pod - https://github.com/jupyterhub/kubespawner/blob/master/kubespawner/proxy.py#L122. I'd love for us to reuse that here instead - although it is possible we can not.